ast.go

```diff
@@ -26,20 +26,101 @@ import (
 // The Serializer interface is implemented by all
 // expressions/statements.
 type Serializer interface {
-	// Serialize writes the statement/expression to the io.Writer. If an
-	// error is returned the io.Writer may contain partial output.
-	Serialize(w io.Writer) error
+	// Serialize writes the statement/expression to the sqlWriter. If an
+	// error is returned the sqlWriter may contain partial output.
+	Serialize(w sqlWriter) error
 }
 
 // Serialize serializes a serializer to a string.
 func Serialize(s Serializer) (string, error) {
 	var buf bytes.Buffer
-	if err := s.Serialize(&buf); err != nil {
+	if err := s.Serialize(&standardWriter{&buf}); err != nil {
 		return "", err
 	}
 	return buf.String(), nil
 }
 
+// SerializeWithPlaceholders serializes a serializer to a string but without substituting
+// values. It may be useful for logging.
+func SerializeWithPlaceholders(s Serializer) (string, error) {
+	var buf bytes.Buffer
+	if err := s.Serialize(&placeholderWriter{&buf}); err != nil {
+		return "", err
+	}
+	return buf.String(), nil
+}
+
+type sqlWriter interface {
```

Comment:
Now that I see this in code, I think this interface should be named `Writer`. External packages would see it as `squalor.Writer` which makes sense.